File: Data Structures & Algorithms/two-integer-sum/submission-3.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # nums[i] + nums[j] = target
        # nums[i] = target - nums[j]
        # Bottleneck of an O(n^2) solution is comparing i and j -> nested for loop
        hmap = {}
        for i in range(len(nums)):
            hmap[nums[i]]=i
        for x in range(len(nums)):
            diff = target-nums[x]
            if diff in hmap and x != hmap[diff]:
                return [x, hmap[diff]]
        return []

File: Data Structures & Algorithms/two-integer-sum-ii/submission-3.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int l = 0, r = numbers.size()-1;
        while (l < r) {
            int sum = numbers[l] + numbers[r];
            if (target == sum) break;
            if (sum > target) r--;
            else l++;
        }
        return {l+1, r+1};
    }
};
